File: src/main/java/com/github/liebharc/queryenrichment/SchemaDocumentation.java
package com.github.liebharc.queryenrichment;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * A helper to draw a simple documentation for the attribute schema
 */
public class SchemaDocumentation {

    /** Singleton */
    public static final SchemaDocumentation INSTANCE = new SchemaDocumentation();

    private SchemaDocumentation() {
        // Singleton class
    }

    public String drawSchema(Collection<Attribute<?>> attributes) {
        final StringBuilder result = new StringBuilder();
        final Map<String, List<Attribute>> byDomain =
                attributes.stream().collect(Collectors.groupingBy(Attribute::getDomain));
        byDomain.keySet().stream().sorted().forEach(domain -> {
            result.append(domain);
            result.append(":");
            result.append("\n");
            byDomain.get(domain).stream().sorted(Comparator.comparing(Attribute::getProperty)).forEach(attr -> {
                result.append("\t");
                result.append(" - ");
                result.append(attr.getProperty());
                result.append(": ");
                result.append(attr.getAttributeClass().getSimpleName());
                result.append("\n");
            });
        });

        return result.toString();
    }
}

File: src/main/java/com/github/liebharc/queryenrichment/SchemaDocumentation.kt
package com.github.liebharc.queryenrichment

/**
 * A helper to draw a simple documentation for the attribute schema
 */
class SchemaDocumentation private constructor()// Singleton class
{

    fun drawSchema(attributes: Collection<Attribute<*>>): String {
        val result = StringBuilder()

        val byDomain = attributes.groupBy { attr -> attr.domain }
        byDomain.keys.stream().sorted().forEach { domain ->
            result.append(domain)
            result.append(":")
            result.append("\n")
            byDomain[domain]?.sortedBy { it.property }?.forEach { attr ->
                result.append("\t")
                result.append(" - ")
                result.append(attr.property)
                result.append(": ")
                result.append(attr.attributeClass.simpleName)
                result.append("\n")
            }
        }

        return result.toString()
    }

    companion object {

        /** Singleton  */
        val INSTANCE = SchemaDocumentation()
    }
}
